docs/api/arg-types.md

```diff
@@ -0,0 +1,511 @@
+    'common/arg-types-options.js.mdx',
+    'common/arg-types-options.ts.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+### `table`
+
+Type:
+
+```ts
+{
+  category?: string;
+  defaultValue?: {
+    detail?: string;
+    summary: string;
+  };
+  subcategory?: string;
+  type?: {
+    detail?: string;
+    summary: string;
+  };
+}
+```
+
+Default: [Inferred](#automatic-argtype-inference)
+
+Specify how the arg is documented in the [`ArgTypes` doc block](./doc-block-argtypes.md), [`Controls` doc block](./doc-block-controls.md), and [Controls addon panel](../essentials/controls.md).
+
+<!-- prettier-ignore-start -->
+
+ <CodeSnippets
+  paths={[
+    'angular/arg-types-table.ts.mdx',
+    'web-components/arg-types-table.js.mdx',
+    'web-components/arg-types-table.ts.mdx',
+    'common/arg-types-table.js.mdx',
+    'common/arg-types-table.ts.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+#### `table.category`
+
+Type: `string`
+
+Default: [Inferred](#automatic-argtype-inference), in some frameworks
+
+Display the argType under a category heading, with the label specified by `category`.
+
+#### `table.defaultValue`
+
+Type: `{ detail?: string; summary: string }`
+
+Default: [Inferred](#automatic-argtype-inference)
+
+The documented default value of the argType. `summary` is typically used for the value itself, while `detail` is used for additional information.
+
+#### `table.subcategory`
+
+Type: `string`
+
+Display the argType under a subcategory heading (which displays under the [`category`] heading), with the label specified by `subcategory`.
+
+#### `table.type`
+
+Type: `{ detail?: string; summary: string }`
+
+Default: [Inferred](#automatic-argtype-inference)
+
+The documented type of the argType. `summary` is typically used for the type itself, while `detail` is used for additional information.
+
+### `type`
+
+Type: `'boolean' | 'function' | 'number' | 'string' | 'symbol' | SBType`
+
+The full type of `SBType` is:
+
+<details>
+<summary>SBType</summary>
+
+```ts
+interface SBBaseType {
+  required?: boolean;
+  raw?: string;
+}
+
+type SBScalarType = SBBaseType & {
+  name: 'boolean' | 'string' | 'number' | 'function' | 'symbol';
+};
+
+type SBArrayType = SBBaseType & {
+  name: 'array';
+  value: SBType;
+};
+type SBObjectType = SBBaseType & {
+  name: 'object';
+  value: Record<string, SBType>;
+};
+type SBEnumType = SBBaseType & {
+  name: 'enum';
+  value: (string | number)[];
+};
+type SBIntersectionType = SBBaseType & {
+  name: 'intersection';
+  value: SBType[];
+};
+type SBUnionType = SBBaseType & {
+  name: 'union';
+  value: SBType[];
+};
+type SBOtherType = SBBaseType & {
+  name: 'other';
+  value: string;
+};
+
+type SBType =
+  | SBScalarType
+  | SBEnumType
+  | SBArrayType
+  | SBObjectType
+  | SBIntersectionType
+  | SBUnionType
+  | SBOtherType;
+```
+
+</details>
+
+Default: [Inferred](#automatic-argtype-inference)
+
+TK <!-- TODO: Not sure how this differs from `table.type.summary` -->
```

Comment:
Highlighting this TODO, for which I need an answer, please.

Reply:
TLDR: The `type` field is semantic and the `table.type.summary` is what's shown in the UI.

It's a little convoluted and I hope we can streamline this in 8.0.
1. The component is analyzed using some kind of docgen (`react-docgen`/`compodoc`/etc.) All of these tools have different formats.
2. Those all get summarized in the `type` field. Users can also specify the `type` field themselves, in case docgen got it wrong, or there is no docgen for your particular framework/renderer.
3. Storybook uses the `type` field to auto-generate the `control` field (which can also be manually overwritten/refined)
4. Storybook also uses the `type` field to autogenerate the `table.type.summary/details` fields for display (which can also be manually overwritten/refined)

There are two ways of specifying type:

```js
// longhand
export default {
  argTypes: { 
     foo: { type: { name: 'number' } },
  }
}
```

```js
// shorthand
export default {
  argTypes: { 
     foo: { type: 'number' },
  }
}
```

Why not just use the shorthand? Well, there is some support for structured types like Object/Array. This was overengineered in retrospect since we don't take advantage of that data yet. We will in the future, but I don't know when. If I could turn back time, I would just use the shorthand.

Also worth noting that we're planning on introducing an even shorter shorthand in 7.2 which will look like:

```js
export default {
  argTypes: {
     foo: 'number',
     bar: 'action', // <= this is new
  } 
}
```

We'll probably create an RFC for this change. The rationale behind it is a bit complex but it has to do with performance, story portability, etc. (making it easier to not have to run docgen at all in some cases). Should not affect the current docs, but just wanted to give the full picture that this will be changing soon.

Reply:
> Why not just use the shorthand? Well, there is some support for structured types like Object/Array. This was overengineered in retrospect since we don't take advantage of that data yet. We will in the future, but I don't know when. If I could turn back time, I would just use the shorthand.

I agree with you @shilman for a long time it was confusing for me, I thought it is a new API. I think we need to normalize and simplify the API, and it should be always an option to override values.

Another important point @shilman, should we separate the docgen plugins, right you are kind of normalizing react and vue, the caveat here is not all renderers have the same architecture

Reply:
Thank you for the detailed answer! 👏

Updates pushed in [55bfceb](https://github.com/storybookjs/storybook/pull/22970/commits/55bfcebd26a4b42f2baee509691fb06004e8bb49).